Custom Fields: Refactor AztecEditorView to be reusable

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

This PR continues the feedback in #13863 to make `AztecEditorView` reusable.

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

With the Custom Fields Editing project, we want to be able to use the Aztec Editor within SwiftUI, as the custom fields editing area are built with it. This is done by creating `AztecEditorView` as `UIViewControllerRepresentable`.

What this PR adds:

1. `AztecEditorView` moved to its own file for reusability.
2. `EditorFactory` updated with `customFieldRichTextEditor()` function and made more generic for non-product contexts.
3. Added `onContentChanged` to retrieve the latest content in Aztec.
4. Adds implementation of the `AztecEditorView` in `CustomFieldEditorView`.
5. Adds logic to check whether there's a recent change in the content of Key / Value field. It is used to enable/disable "Save" button.

In hindsight, the implementation could probably be better in a separate PR, but I hope this is small enough to check both.

| Text | HTML |
|-|-|
| ![Simulator Screenshot - iPhone 15 Pro - 2024-09-24 at 14 53 07](https://github.com/user-attachments/assets/2ca97a32-0a6a-4cb2-ad4c-a641ba254367) |  ![Simulator Screenshot - iPhone 15 Pro - 2024-09-24 at 14 53 10](https://github.com/user-attachments/assets/20618bb2-542f-4385-9746-f624eb380769) |


## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Go to Orders and open an Order that has custom fields in it,
2. Tap View Custom Fields, ensure the Custom Fields list screen is shown,
3. Tap a custom field, ensure the editor is loaded with the key and value shown properly
4. **[Testing related to this PR starts here]** Use the "Text"/"HTML" toggle and switch to "HTML"
6. Ensure that the Aztec Editor is loaded with the value of the custom field shown
8. Add some changes to the value, ensure the "Save" button is now enabled, 
9. Switch back to "Text",
10. Ensure that the Text Editor shows the latest value,
11. Do the same Aztec test with other custom fields that has HTML string value. Aztec should render the HTML correctly.

** Known behavior / possible issue **
- When switching from Text to HTML, Aztec will automatically try to format things the way it wants and it will trigger a value change. You can see this in the video below where the `<p>` tag is automatically stripped. This is still being discussed and if any change is needed, it will be handled in a separate PR. Feel free to ignore for now.
- Some strings are not translatable yet, these are functionality-related strings and I'm keeping them that way to keep it flexible in case we need to change them. They will be updated in later PRs.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

I tested this in Simulator for iPhone with iOS 17.5. Check video below for more details.

As suggested [downthread](https://github.com/woocommerce/woocommerce-ios/pull/13925#pullrequestreview-2324650489), I also smoke-tested Aztec editor in Product Description and Short Description and they worked fine in my test. Check video below also.

## Video
<!-- Include before and after images or gifs when appropriate. -->

Aztec Editor:

https://github.com/user-attachments/assets/049533e4-9d46-4d47-9e75-e41ac942e44e

Smoke tests in Products:


https://github.com/user-attachments/assets/25753e80-4de8-4ed9-adc8-4f32aedd345c


## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
Refactor EditorFactory to make it easier to create editor used in non-product.

hafizrahman committed:
Re-add passing product to view controller and simple content parameter refactor.

hafizrahman committed:
Update onContentSave

hafizrahman committed:
Add factory function for generating custom fields rich text editor.

hafizrahman committed:
Put AztecEditorView into its own file so it's reusable in the future

hafizrahman committed:
Try: Make editor's Done button work by passing the latest change to the regular text editor.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/226">20.2.1</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add comment to explain AztecEditorView usage.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Editor/AztecEditorView.swift:
```diff
@@ -0,0 +1,21 @@
+///
+struct AztecEditorView: UIViewControllerRepresentable {
+    @Binding var value: String
+    @Binding var viewController: AztecEditorViewController?
```
As discussed on Slack, it's hacky to set the view controller as `@Binding` because bindings are supposed to be used for properties that are subject to changes.

Instead, you should get the value of the html content from the editor and set it to the `value` binding. My suggestion:
- Create a coordinator for the AztecEditorView.
- Set the delegate for `htmlTextView` of `AztecEditorViewController.editorView` to this coordinator.
- In the `textViewDidChange` method of the `UITextViewDelegate` conformance, set the `text` binding to the updated text similar to [this](https://github.com/shaps80/SwiftUIBackports/blob/f5f23b016eeda6642a0fe1020241af19c9c05556/Sources/SwiftUIBackports/iOS/TextEditor/TextEditor.swift#L130).

Let me know if you have any questions.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Editor/AztecEditorView.swift:
```diff
@@ -0,0 +1,21 @@
+    @Binding var viewController: AztecEditorViewController?
+
+    func makeUIViewController(context: Context) -> AztecEditorViewController {
+        let controller = EditorFactory().customFieldRichTextEditor(initialValue: value) as! AztecEditorViewController
```
Please avoid using forced unwrapping whenever possible. Instead, add a guard check for the type and trigger `fatalError` with a clear message to make it easier to debug.

hafizrahman committed:
Revert "Try: Make editor's Done button work by passing the latest change to the regular text editor."

This reverts commit 53494ce7bc3a50becff83007819eec22e8cd2cce.

hafizrahman committed:
Update AztecEditorView by wrapping a navigation bar to Aztec so its navigation bar appears. Also add optional Cancel button.

hafizrahman committed:
Add `onDone` parameter for callback when Aztec's done button is tapped.

hafizrahman committed:
Update Custom List View's cancel button.

hafizrahman committed:
Fix typo.

itsmeichigo commented:
@hafizrahman Is this PR ready for another review yet?

hafizrahman commented:
@itsmeichigo thanks for the ping! I left this PR this week for HACK week, but I'll get back to you on this on Monday. At the very least I still need to update this with the latest `trunk`

hafizrahman committed:
Merge branch 'trunk' into feat/13493-refactor-aztec-make-more-generic

hafizrahman committed:
Revert "Fix typo."

This reverts commit f32a3f3636e4aa55cf4c9f2d7753cc6671913a36.

hafizrahman committed:
Revert "Update Custom List View's cancel button."

This reverts commit 6d5855b0786e755591ec7e391e6456757251ff56.

hafizrahman committed:
Revert "Add `onDone` parameter for callback when Aztec's done button is tapped."

This reverts commit 9fd2f57d4cbe8a6bd30ccab992f32349361b62dc.

hafizrahman committed:
Revert "Update AztecEditorView by wrapping a navigation bar to Aztec so its navigation bar appears. Also add optional Cancel button."

This reverts commit 60b44a4ad9902e232c04aa8ee47943171e5cb410.

hafizrahman committed:
Add onContentChanged optional callback to Aztec Editor.

hafizrahman committed:
Update UI to use segmented control to switch between Text and HTML

hafizrahman committed:
Add onContentChanged initialization to AztecEditorView

hafizrahman committed:
Implement onContentChanged in the View.

hafizrahman committed:
Handle "Modifying state during view update" issue

hafizrahman committed:
Refactor hasUnsavedChanges logic

Also fixes "Modifying state during view update" in AztecEditorView

hafizrahman committed:
Update UI to put the Picker on right side of Value label

hafizrahman committed:
Input text styling

hafizrahman committed:
Update logic to strip html in list view but not in editor view

hafizrahman committed:
Refactor AztecEditorView remove unneeded binding and update documentation.

hafizrahman committed:
Remove irrelevant unit test.

HTML should not be stripped now since they need stay there for editing purposes.

hafizrahman commented:
@itsmeichigo We had a last minute design change yesterday so I had to rework everything. This should be good to review again.

I have updated the original post with the latest instruction and screenshots/video, so perhaps it's easiest to treat this as a new review.

In terms of reviewing by commits, I had a few revert commits to undo the previous work, and the latest ones to review are the commits starting from 31cf7e388cf7dc24037813a457d1c817b927b180

hafizrahman committed:
Make some strings translatable.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorView.swift:
```diff
@@ -1,39 +1,18 @@
     @State private var key: String
     @State private var value: String
-    @State private var hasUnsavedChanges: Bool = false
     @State private var showRichTextEditor: Bool = false
     @State private var showingActionSheet: Bool = false
```
Super nit:
```suggestion
    @State private var showRichTextEditor = false
    @State private var showingActionSheet = false
```

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorView.swift:
```diff
@@ -53,54 +32,64 @@ struct CustomFieldEditorView: View {
-                            .buttonStyle(.plain)
-                            .foregroundColor(Color(uiColor: .accent))
+                            .pickerStyle(.segmented)
+                            .fixedSize()
```
❓ Any reason to use `fixedSize()` here? From my experience, this modifier brings more issues than fixes - it can cause unexpected layout behaviors like stretched views.

hafizrahman replied:
without `.fixedSize()`, `.segmented` picker style by default stretches itself to all available width:
<img src="https://github.com/user-attachments/assets/c926406f-b11f-465f-8ec1-de12de1c357f" width="42%" />

This looks especially bad on wider screen:

<img src="https://github.com/user-attachments/assets/239e193a-4498-464e-b488-9d64876ffe40" width="42%" />

itsmeichigo replied:
You can make the Picker smaller by letting the "Value" text have the maximum width possible with `.frame(maxWidth: .infinity, alignment: .leading)`.

hafizrahman replied:
Thanks, that's a good solution. https://github.com/woocommerce/woocommerce-ios/pull/13925/commits/fca268c1a8d5570dfdb2c98858853897166bde67

I wasn't aware of the issue with `fixedSize()`, thanks for mentioning that.

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorView.swift:
```diff
@@ -53,54 +32,64 @@ struct CustomFieldEditorView: View {
+                    if showRichTextEditor {
+                        AztecEditorView(initialValue: value,
+                                        onContentChanged: { content in
+                            DispatchQueue.main.async {
+                                value = content
+                            }
```
❓ Could you explain the need for using `DispatchQueue.main.async` here?

hafizrahman replied:
Ah, I added this in the commit message but it wasn't obvious:

> Also fixes "Modifying state during view update" in AztecEditorView

This fixes the same issue as in the 314a522870ec1f953109b9cd0d5350d165e9d9ff commit.

itsmeichigo replied:
That error hints an incorrect use of SwiftUI. Using `DispatchQueue.main.async` is just a workaround, and I'm unsure it actually fixes the problem.

Instead, try using `@Binding` only:

```Swift
struct AztecEditorView: UIViewControllerRepresentable {
    @Binding var content: String

    func makeUIViewController(context: Context) -> AztecEditorViewController {
        let controller = EditorFactory().customFieldRichTextEditor(initialValue: content)

        guard let aztecController = controller as? AztecEditorViewController else {
            fatalError("EditorFactory must return an AztecEditorViewController, but returned \(type(of: controller))")
        }

        aztecController.onContentChanged = { content in
            self.content = content
        }
        return aztecController
    }
}
```

hafizrahman replied:
Thanks, that solution makes sense to me but it's showing the same issue:

![Screenshot 2024-09-25 at 21 54 23](https://github.com/user-attachments/assets/484e2821-897a-4072-bf36-3b9471b79410)

(The warning only appears when actually tapping the "HTML" switcher in the Simulator).

itsmeichigo replied:
This happens because `content` is updated immediately upon view creation. A workaround:

<img width="940" alt="image" src="https://github.com/user-attachments/assets/268c2ef7-ca6f-4575-8bba-a3d68ab4f98b">

hafizrahman replied:
Updated in https://github.com/woocommerce/woocommerce-ios/pull/13925/commits/60d5f4b2c8f34738b34c54cfbb4539e829a1828d

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorView.swift:
```diff
@@ -53,54 +32,64 @@ struct CustomFieldEditorView: View {
+                            .subheadlineStyle()
+                            .frame(minHeight: Layout.minimumEditorSize)
+                            .padding(insets: Layout.inputInsets)
+                            .background(Color(.listForeground(modal: false)))
```
Since there are rounded corners on the overlay, the background view should have rounded corners as well. If you look closely, the background view is currently showing underneath which doesn't look nice.

<img width="134" alt="image" src="https://github.com/user-attachments/assets/bf99f566-b73a-45d5-8451-fd7432f3a9ce">

Same comment for the other fields.

hafizrahman replied:
Wow, great eye! So these fields are pretty much copied from `SupportForm` (Settings > Help > Contact Support) as they're the only View I could find that have more than one text input fields in one screen in the app. 

I see that it also has the same problem:
<img width="438" alt="Screenshot 2024-09-24 at 18 55 18" src="https://github.com/user-attachments/assets/af22b622-7baa-424b-a501-69586f03dbfd">

I'll handle this in a separate issue to fix both views at once.

hafizrahman replied:
#14028

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Editor/AztecEditorView.swift:
```diff
@@ -0,0 +1,26 @@
+        }
+
+        aztecController.onContentChanged = onContentChanged
+        return controller as! AztecEditorViewController
```
Since you already use the guard check above, can you return `aztecController` here instead of `controller` with forced unwrapping?

hafizrahman replied:
Oh huh this is definitely not intended to be there. I fixed this then I think it returned after one of the reverts and I forgot to update again.  Updated in https://github.com/woocommerce/woocommerce-ios/pull/13925/commits/adeb80deb0da453b447b1e01278ea50d00fe3c3e

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Custom Fields/CustomFieldEditorView.swift:
```diff
@@ -53,54 +32,64 @@ struct CustomFieldEditorView: View {
-                            checkForModifications()
-                        }
+                    } else {
+                        TextEditor(text: isReadOnlyValue ? .constant(value) : $value)
```
Since there's a placeholder text on both the key field and the HTML mode, should we add placeholder text for the text mode as well?

hafizrahman replied:
So unlike in Key field where having a value is required (although the checking logic is not yet added here), having empty content in Value field is a supported feature of Custom Fields. In this case, we want the empty content to be displayed as-is, instead of displaying a placeholder message.

This is also the same as in Core:

<img width="654" alt="Screenshot 2024-09-24 at 19 49 36" src="https://github.com/user-attachments/assets/67bc5d8f-837c-4da5-9b66-46c9818cea94">

itsmeichigo reviewed (commented):
Thanks for the updates @hafizrahman. Since you updated some code in `EditorFactory` affecting Product description and short description, it's worth it to do smoke test on the features to ensure they still work as expected.

I left some other comments for improvements below.

itsmeichigo commented:
Also, I tried [this sample HTML code](https://gist.github.com/chrisvfritz/bc010e6ed25b802da7eb) in the HTML mode but the body content does not show up. Is that an issue with Aztec, should we raise an issue for the library?


https://github.com/user-attachments/assets/c84602c7-4947-44ab-8c44-93bf7d1d8dcd

hafizrahman committed:
Simplify vars

Co-authored-by: Huong Do <hi.huongdo@gmail.com>

hafizrahman commented:
> Also, I tried [this sample HTML code](https://gist.github.com/chrisvfritz/bc010e6ed25b802da7eb) in the HTML mode but the body content does not show up. Is that an issue with Aztec, should we raise an issue for the library?

That is indeed Aztec's default behavior, it:
- only accepts certain subset of HTML tags, and others are stripped off.
- some tags are changed to square bracket/shortcode format.

Here's the same behavior from Short Descriptions in Products. I'm using the Code icon toggle in the toolbar to switch from code to rich text mode, and the autoformatting happened there too:


https://github.com/user-attachments/assets/58f8f6be-58d4-4b71-8296-456e27c87f64


This seems to stem from the fact that Aztec is meant for writing blog posts, thus tags like `html` or `body` should be stripped off as it would create issue in a blog post context.

We discussed this in Slack p1725529996823859/1724862433.885169-slack-C03L1NF1EA3 . Initially we wanted to use Aztec immediately if HTML content is detected, but the issue you saw made us change it so that merchants have to intentionally choose HTML editor for this to happen. Our hypothesis also is that HTML editing is going to be a small subset of the usage, and it's more like a nice-to-have feature. This is just the i1 decision though, and we should listen to users for future iterations.

hafizrahman committed:
Remove forced casting

hafizrahman commented:
Thanks for your reviews @itsmeichigo , this is good for another round.

Update: I also updated the original post with my smoke testing Aztec on Product. Thank you for the suggestion!

itsmeichigo commented:
I left some responses in the discussions above, could you take a look? They were not set as a new round of review.

hafizrahman committed:
Avoid using fixedSize() and update the label's appearance instead to achieve the same effect.

hafizrahman committed:
Refactor the way the content in Aztec is bind to the value state in the view controller.

This helps avoid the  "Modifying state during view update, this will cause undefined behavior" issue.

hafizrahman committed:
Merge branch 'trunk' into feat/13493-refactor-aztec-make-more-generic

hafizrahman commented:
@itsmeichigo This is good for another review, thanks. Do let me know if I miss something.

itsmeichigo reviewed (approved):
Thanks for the updates!
